src/codecs/openexr.rs

```diff
@@ -0,0 +1,433 @@
+//! Decoding of OpenEXR (.exr) Images
+//!
+//! OpenEXR is an image format that is widely used, especially in VFX,
+//! because it supports lossless and lossy compression for float data.
+//!
+//! This decoder only supports RGB and RGBA images.
+//! If an image does not contain alpha information,
+//! it is defaulted to `1.0` (no transparency).
+//!
+//! # Related Links
+//! * <https://www.openexr.com/documentation.html> - The OpenEXR reference.
+
+// ROADMAP: See https://github.com/image-rs/image/pull/1475
+
+extern crate exr;
+use exr::prelude::*;
+
+use crate::{ImageDecoder, ImageResult, ColorType, Progress, ImageError, ImageFormat, ImageBuffer, Rgba, Rgb, ImageEncoder, ExtendedColorType};
+use std::io::{Write, Seek, BufRead, Cursor, BufReader, BufWriter};
+use crate::error::{DecodingError, ImageFormatHint, LimitError, LimitErrorKind, EncodingError};
+use crate::image::decoder_to_vec;
+use std::path::Path;
+use crate::buffer_::{Rgb32FImage, Rgba32FImage};
+
+
+// TODO this could be a generic function that works for any format
```

Comment:
This (and the other functions) have the same purpose as `ImageBuffer::save<Q>(path: Q)`, but with two differences:
1. It works without having to implement `Pixel::COLOR_TYPE` for `Rgba<f32>`.
2. There is an additional version that works with generic readers and writers (`impl Read`), not only one version with a path

Instead of increasing the API surface of the OpenEXR module, we could have them as `pub(crate)` and wait until the `Pixel` trait is implemented for `Rgb<f32>`, and furthermore add the non-path variants to the `ImageBuffer` type.

Reply:
This increases the surface quite a bit. Note that for (2) specifically there is the [`image::io::Reader`](https://docs.rs/image/0.23.14/image/io/index.html) which allows a composition of the different modes without a multiplicative effect on the number of methods. The raw methods are slated for deprecation once the equivalent writer exists.

Reply:
Sorry if I didn't understand that, english is not my mother tongue. Did I get that right that you suggest I keep these methods and later mark them deprecated once `Rgb<f32>` implements `Pixel` and dynamic image supports `Rgba32F`?

Reply:
@HeroicKatora So, do we keep these?

Reply:
Could you replace these with new `Reader::{decode_rgb32f, decode_rgba32f}` methods?

Reply:
Sure, sounds awesome, where can I find those methods? Are they in the master branch? Couldn't find them, not even with global search in the whole `image` repository lol

Reply:
They don't exist now. I was suggesting you add them to the [`io::Reader`](https://docs.rs/image/0.23.14/image/io/struct.Reader.html) struct.

Reply:
Oh, haha :D Yes I'd love to do that, now I got it

Reply:
Are we going to add the other variants later? Like `Rgb8` and others?

Reply:
Update: Moving these methods turns out to be a difficult refactor. It seems to be too specific for the general reader interface. I'm stuck with the following code:

```rust

    fn read_as_buffer<Pixel, PixelComponent, PixelStorage>(self) -> ImageResult<ImageBuffer<Pixel, PixelStorage>> {
        let decoder: Box<dyn ImageDecoder> = ???;
        let (width, height) = decoder.dimensions();
        let buffer: Vec<PixelComponent> = decoder_to_vec(decoder)?;
        ImageBuffer::from_raw(width, height, buffer)
    }
```

Maybe this should rather be a function with the following type:

```rust
trait Decoder {
    // ...

    fn decode_as_buffer <Px: Pixel> (self) -> ImageBuffer<Px, Vec<Px::SubPixel>> {
        assert_eq!(Px::COLOR_TYPE, self.color_type());
        let (width, height) = self.dimensions();
        let buffer: Vec<Px::SubPixel> = decoder_to_vec(self)?;
        ImageBuffer::from_raw(width, height, buffer)
    }
}
```

Reply:
I don't quite follow, where does the necessity of all the type parameters come from? In the implementation here you just had `ImageResult<Rgba32FImage>` which should be fine. In particular there is no need to generalize to _all_ buffer types and to _all_ `impl Pixel` types. Rather, and this might solve your confusion on `Box<dyn ImageDecoder>`, I wouldn't generalize this further than for pixel types whose components implement `bytemuck::Pod`. This covers all integers and floats which should be more then fine.

This allows you to decode to a vector that you're allocating yourself by using something like: `ImageDecoder::read_image(bytemuck::cast_slice_mut(&mut the_sample_vec[..]))`

Reply:
> This allows you to decode to a vector that you're allocating yourself by using something like: ImageDecoder::read_image(bytemuck::cast_slice_mut(&mut the_sample_vec[..]))

The only point of these helper methods is to allocate a buffer, such that I don't have to repeat the boilerplate allocation code twice in every test. This is less error prone and the intent is immediately obvious, resulting in cleaner code. Maybe this simpler method could even be used in other places in the existing `image` crate code.

---
 
> I wouldn't generalize this further than for pixel types whose components implement bytemuck::Pod. This covers all integers and floats which should be more then fine.

Yes, that's what I did:

```rust

trait ImageDecoder {
    // ...

    /// Fills an `ImageBuffer` with the image pixels.
    /// You must provide a pixel type that exactly matches the image color type,
    /// no conversion is done. For incompatible pixel types, an error is returned.
    fn decode_into_pixels<Px: Pixel> (self) -> ImageResult<ImageBuffer<Px, Vec<Px::Subpixel>>>
        where Px::Subpixel: bytemuck::Pod,
              Px: 'static,
    {
        if Px::COLOR_TYPE != self.color_type() {
            return Err(ImageError::Decoding(DecodingError::new(
                ImageFormatHint::Unknown, // TODO we may know this
                "The color type of the provided Pixel must match the color type of the ImageDecoder"
            )));
        }

        let (width, height) = self.dimensions();
        let buffer: Vec<Px::Subpixel> = decoder_to_vec(self)?;

        ImageBuffer::from_raw(width, height, buffer)

            // this should be the only reason for the "from raw" call to fail,
            // even though such a large allocation would probably cause an error much earlier
            .ok_or_else(|| ImageError::Limits(LimitError::from_kind(LimitErrorKind::InsufficientMemory)))
    }
}


fn usage(){
    OpenExrDecoder::read(read, Some(true))?.decode_into_pixels::<Rgba<f32>>()
}
```


I got this working, but the Pixel trait is implemented incorrectly for `Rgb<f32>`. The `Rgb<f32>::SubPixel` type will not resolve to an `f32` color type. The Pixel trait is only really implemented for `Rgb8` and `Rgb16`, not `f32`. This causes the error check in the first line to always fail for `f32` colors.


Instead of this generic implementation, we could copy & paste all the implementations manually for all the color types.... Is this really the way to go? The pixel type must be fixed sooner or later anyways, so maybe we should probably just do that instead of manually copying the code.

Reply:
> The only point of these helper methods is to allocate a buffer, such that I don't have to repeat the boilerplate allocation code twice in every test. 
> The Rgb<f32>::SubPixel type will not resolve to an f32 color type. The Pixel trait is only really implemented for Rgb8 and Rgb16, not f32

This helper method as sketched would then be better suited for `impl ImageBuffer`. It is similar to [`DynamicImage::from_decoder`](https://docs.rs/image/0.23.14/image/enum.DynamicImage.html#method.from_decoder). I'm afraid the `Pixel` issue is something we can't easily resolve in this major version of `image`. We could, however, already add an improved trait that is _just_ used for these method. Something like:

```
#[repr(u8)]
enum SampleType {
    U8 = 0,
    U16 = 1,
…
    F32 = 2,
}

trait Sample: Primitive {
    const TYPE: SampleType;
}

mod hidden {
    pub trait PixelV2 { …  }
    impl<T: Sample> PixelV2 for Rgb<T> { … }
}

impl<Px> ImageBuffer<Px, Vec<Px::Subpixel>>
where
    Px: hidden::PixelV2,
{
    fn from_decoder(mut decoder: impl ImageDecoder) -> ImageResult<Self> {
        if Px::COLOR_TYPE != decoder.color_type() { … }
    }
}
```

In any case, I would much rather discuss these helpers in a separate PR _after_ the core EXR functionality has been added together with the new `ColorType` variant :)

Reply:
Interesting idea, I like it. I propose we keep the helper methods, but private, then we can delay the discussion without duplicating code in the tests. Okay?

Reply:
I'm fine with private helpers :+1: